subprojects/jvm-services/src/main/java/org/gradle/jvm/toolchain/internal/MavenToolchainsInstallationSupplier.java

```diff
@@ -84,10 +87,11 @@ protected Set<InstallationLocation> findCandidates() {
                     .map(jdkHome -> new InstallationLocation(new File(jdkHome), "Maven Toolchains"))
                     .collect(Collectors.toSet());
             } catch (IOException | ParserConfigurationException | SAXException | XPathExpressionException e) {
+                String toolchainValue = toolchainLocation.getOrElse("?");
                 if (LOGGER.isDebugEnabled()) {
-                    LOGGER.debug(String.format("Java Toolchain auto-detection failed to parse Maven Toolchains located at %s", toolchainLocation), e);
+                    LOGGER.debug("Java Toolchain auto-detection failed to parse Maven Toolchains located at {}", toolchainValue, e);
```

Comment:
I believe we can log the `toolchainFile` directly which removes the need for the `getOrElse` above. Though we might need to explicitly get the absolute path.

Reply:
We were running into this problem in the issue this PR responds to where instead of the value being logged, we get `provider(?)`. The toString is implemented here, which shows only the type of the `Provider` is printed in the `toString`:
https://github.com/gradle/gradle/blob/master/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/AbstractMinimalProvider.java#L163-L165

Reply:
Missed this completely. Removed the extra provider extraction, and logged `toolchainFile` directly.